rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,718 @@
+to be permanently unhealthy since it is never able to initialize loading the affected resource.
+
+#### Delete
+
+The `Delete` RPC takes the parameters required to match a resource and performs a hard delete of the specified resource
+from the backend and returns a `google.protobuf.Empty`.
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Remove a matching Foo resource
+    rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);
+
+message DeleteFooRequest {
+  // Name of the foo to remove. Some resource may require more parameters to match and
+  // may not use the name at all.
+  string foo_id = 1;
+}
+```
+
+### Backend Storage
+
+A backend service to handle persisting and retrieving a resource from the backend is typically defined in
+`lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.
+Continuing on with the example above, the sections below show how the backend service for the `Foo` resource might look like.
+
+The sections also contain a reference example for how to interact with the backend to perform common operations on a
+resource. For most cases, when adding a new resource, it is preferred to create a service that wraps
+the [generic.Service](https://github.com/gravitational/teleport/blob/7f3c58df1fd675a813dc2992c10b2796b9b5c6bf/lib/services/local/generic/generic.go#L73-L81)
+over implementing everything from scratch. If custom behavior is required for a subset of backend operations, they may
+be implemented directly while all the other operations still make use of the generic service.
+
+#### Resource validation
+The strictest validation of a resource should be performed prior to write operations. Any resource persisted in the
+backend should be guaranteed to be valid. Read operations should not perform resource validations, doing so could
+prevent a resource being read if validations are modified to be more restrictive after a resource had already been
+written. This allows for resilient reads which ensure that any resource stored is always allowed to be retrieved from
+the backend.
+
+#### Create
+
+When creating a new resource, the `backend.Backend.Create` method should be used to persist the resource. It is also
+imperative that the revision generated by the backend is set on the returned resource.
+
+```go
+func (s *FooService) CreateFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
+	value, err := convertFooToValue(foo)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	item := backend.Item{
+		Key:     backend.Key("foo", foo.GetName()),
+		Value:   value,
+		Expires: foo.Expiry(),
+	}
+
+	lease, err := s.backend.Create(ctx, item)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Update the foo with the revision generated by the backend during the write operation.
+	foo.GetMetadata().SetRevision(lease.Revision)
```

Comment:
This isn't done in `CreateResource` in `service.Generic` on master: https://github.com/gravitational/teleport/blob/1dcd7b37a673eb813245ea041a18e934f6061bea/lib/services/local/generic/generic.go#L192C4-L205

```go
// CreateResource creates a new resource.
func (s *Service[T]) CreateResource(ctx context.Context, resource T) error {
        item, err := s.MakeBackendItem(resource, resource.GetName())
        if err != nil {
                return trace.Wrap(err)
        }


        _, err = s.backend.Create(ctx, item)
        if trace.IsAlreadyExists(err) {
                return trace.AlreadyExists("%s %q already exists", s.resourceKind, resource.GetName())
        }


        return trace.Wrap(err)
}
```

Should it be updated?